dla_cnn/desi/preprocess.py

```diff
@@ -92,3 +92,68 @@ def label_sightline(sightline, kernel, REST_RANGE, pos_sample_kernel_percent=0.3
     # classification is 1 / 0 / -1 for DLA/nonDLA/border
     # offsets_array is offset
     return classification, offsets_array, column_density
+
+def rebin(sightline, v = 20000):
+    """
+    Resample and rebin the input Sightline object's data to a constant dlambda/lambda dispersion.
+
+    Parameters
+    ----------
+    sightline: dla_cnn.data_model.Sightline
+    v: float
+    
+    Returns
+    -------
+    sightline: dla_cnn.data_model.Sightline
+    """
+    c = 2.9979246e8
+    dlambda = np.log(1+v/c)
+    wavelength = np.exp(np.log(10)*sightline.loglam)
+    central_wavelength = [item.central_wavelength for item in sightline.dlas]
+    max_wavelength = wavelength[-1]/min(central_wavelength)
+    min_wavelength = wavelength[0]/max(central_wavelength)
+    pixels_number = int(np.round(np.log(max_wavelength/min_wavelength)/dlambda))+1
+    new_wavelength = wavelength[0]*np.exp(dlambda*np.arange(pixels_number))
+    
+    npix = len(wavelength)
+    wvh = (wavelength + np.roll(wavelength, -1)) / 2.
+    wvh[npix - 1] = wavelength[npix - 1] + \
+                    (wavelength[npix - 1] - wavelength[npix - 2]) / 2.
+    dwv = wvh - np.roll(wvh, 1)
+    dwv[0] = 2 * (wvh[0] - wavelength[0])
+    med_dwv = np.median(dwv)
+    
+    cumsum = np.cumsum(sightline.flux * dwv)
+    cumvar = np.cumsum(sightline.error * dwv, dtype=np.float64)
+    
+    fcum = interp1d(wvh, cumsum,bounds_error=False)
+    fvar = interp1d(wvh, cumvar,bounds_error=False)
+    
+    nnew = len(new_wavelength)
+    nwvh = (new_wavelength + np.roll(new_wavelength, -1)) / 2.
+    nwvh[nnew - 1] = new_wavelength[nnew - 1] + \
+                     (new_wavelength[nnew - 1] - new_wavelength[nnew - 2]) / 2.
+    
+    bwv = np.zeros(nnew + 1)
+    bwv[0] = new_wavelength[0] - (new_wavelength[1] - new_wavelength[0]) / 2.
+    bwv[1:] = nwvh
+    
+    newcum = fcum(bwv)
+    newvar = fvar(bwv)
+    
+    new_fx = (np.roll(newcum, -1) - newcum)[:-1]
+    new_var = (np.roll(newvar, -1) - newvar)[:-1]
+    
+    # Normalize (preserve counts and flambda)
+    new_dwv = bwv - np.roll(bwv, 1)
+    new_fx = new_fx / new_dwv[1:]
+    # Preserve S/N (crudely)
+    med_newdwv = np.median(new_dwv)
+    new_var = new_var / (med_newdwv/med_dwv) / new_dwv[1:]
+    
+    test = (~np.isnan(new_fx))|(~np.isnan(new_var))
```

Comment:
I like testing for `nan` but there shouldn't be any.
So if there are, have the code `raise` an error.

Reply:
The `nan` is due to the `scipy.interp1d` , for those points out of its interpolation range, it automatically fills this value with nan, so I just use `test` to exclude the `nan`. Maybe there is a better solution, I will have a try.

Reply:
I see.  These should only happen at the very edges of the spectrum.
Best to confirm that is true and then it is ok to trim them off.  I just didn't
want us to lose pixels in the middle of the spectrum.  That would be a problem.

Reply:
I have confirmed these only happen at the very edges of the spectrum. The function will `raise` an error when this is not true. Thanks for your advice!